### ISSUE TITLE
copyto!() between cpu and gpu with subarrays

### ISSUE THREAD

**Issue by btmit** [JuliaGPU/CUDA.jl#491]
Copying a view from the CPU to existing memory on the GPU produces a scalar index error. Copy with the same way CPU-to-CPU and GPU-to-GPU still works.

```
using CUDA
CUDA.allowscalar(false)

a = zeros(ComplexF32, 4,3,4)
b = CUDA.zeros(ComplexF32, 4,3)

copyto!(b, view(a,:,:,1))
```

>ERROR: scalar setindex! is disallowed

If I try to do the copy with a broadcast

```
b .= view(a,:,:,1)
```
I get
>ERROR: GPU compilation of kernel broadcast_kernel(CUDA.CuKernelContext, CuDeviceArray{Complex{Float32},2,1}, Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(identity),Tuple{Base.Broadcast.Extruded{SubArray{Complex{Float32},2,Array{Complex{Float32},3},Tuple{Base.Slice{Base.OneTo{Int64}},Base.Slice{Base.OneTo{Int64}},Int64},true},Tuple{Bool,Bool},Tuple{Int64,Int64}}}}, Int64) failed
KernelError: passing and using non-bitstype argument

**Comment by maleadt** [comment]
> `b .= view(a,:,:,1)`

This is not supported, you can only have GPU arrays participate in a broadcast. So no CPU array, let alone a CPU view.

> copyto!(b, view(a,:,:,1))

This is on purpose, too. We currently only support copying from `Array`, because there's no easy way to dispatch on dense CPU arrays. There's `DenseArray`, but that isn't limited to `A<:Array`:

```
julia> AbstractGPUArray <: DenseArray
true
```

And introducing a `DenseCPUArray` alias like https://github.com/JuliaGPU/CUDA.jl/blob/62ae0c3f9e8298d87c1c2d3a3bfd7401005d9092/src/array.jl#L148-L164 would be too hard on the compiler (the current `DenseCuArray` already has significantly regressed precompilation and load times).

TLDR:

```
julia> CuArray(view(a,:,:,1))
4×3 CuArray{Complex{Float32},2}:
 0.0+0.0im  0.0+0.0im  0.0+0.0im
 0.0+0.0im  0.0+0.0im  0.0+0.0im
 0.0+0.0im  0.0+0.0im  0.0+0.0im
 0.0+0.0im  0.0+0.0im  0.0+0.0im
```

**Comment by btmit** [comment]
Thank you for the information and background.

Is there any way then to copy from a slice of an array on the CPU to pre-allocated memory on the GPU?

**Comment by maleadt** [comment]
You can always `unsafe_copyto!(pointer(dst), pointer(src), n)`, but that's a litttle risky of course.

**Comment by ejmeitz** [comment]
Is there a way to copy from GPU into a CPU view?
